### PR TITLE
More polish for running locally

### PR DESCRIPTION
**Pull request by dsyer**
I can run on my dev box and also in a container (with CI
env var set to nonempty). The riff tests always time out, but
my own integration tests look clean.

### CODE REVIEW

File: macros/invoke_knative_deployer.sh [hunk 1]
```diff
@@ -4,6 +4,7 @@ name=$1
 curl_opts=$2
 expected_data=$3
 
+NAMESPACE=${NAMESPACE:-fats}
```
Comment (scothis) [inline macros/invoke_knative_deployer.sh]: rm

NAMESPACE is part of the FATS contract. We should not define it here. The script should fail if the value is not set.

File: .github/workflows/cleanup.sh [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 set -o nounset
 
+test -z ${FATS_DIR-} && FATS_DIR=`dirname "${BASH_SOURCE[0]}"`/../..
```
Comment (scothis) [inline .github/workflows/cleanup.sh]: rm

this script should never be invoked by anything other than the CI workflow.

File: .github/workflows/cleanup.sh [hunk 3]
```diff
@@ -20,4 +21,6 @@ kubectl delete customresourcedefinitions.apiextensions.k8s.io -l app.kubernetes.
 
 source ${FATS_DIR}/macros/helm-reset.sh
 
+NAMESPACE=${NAMESPACE-fats}
```
Comment (scothis) [inline .github/workflows/cleanup.sh]: rm

this script should never be invoked by anything other than the CI workflow.

File: .github/workflows/install.sh [hunk 4]
```diff
@@ -4,6 +4,7 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+test -z ${FATS_DIR-} && FATS_DIR=`dirname "${BASH_SOURCE[0]}"`/../..
```
Comment (scothis) [inline .github/workflows/install.sh]: rm

this script should never be invoked by anything other than the CI workflow.

File: .github/workflows/run.sh [hunk 5]
```diff
@@ -4,12 +4,16 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+test -z ${FATS_DIR-} && FATS_DIR=`dirname "${BASH_SOURCE[0]}"`/../..
```
Comment (scothis) [inline .github/workflows/run.sh]: rm

this script should never be invoked by anything other than the CI workflow.

File: .github/workflows/run.sh [hunk 6]
```diff
@@ -4,12 +4,16 @@ set -o errexit
...
 source ${FATS_DIR}/.configure.sh
 
 # install tools
 ${FATS_DIR}/install.sh riff
 ${FATS_DIR}/install.sh kubectl
 
+NAMESPACE=${NAMESPACE-fats}
```
Comment (scothis) [inline .github/workflows/run.sh]: rm

this script should never be invoked by anything other than the CI workflow.

File: cleanup.sh [hunk 7]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -o nounset
```
Comment (scothis) [inline cleanup.sh]: 👍

File: clusters/kind/configure.sh [hunk 8]
```diff
@@ -20,6 +20,7 @@ post_registry_start() {
     local registry_ip=$(docker inspect --format "{{.NetworkSettings.IPAddress }}" registry)
 
     # patch /etc/containerd/config.toml
+    rm -f containerd-config-patched.toml
```
Comment (scothis) [inline clusters/kind/configure.sh]: 👍

File: clusters/kind/start.sh [hunk 9]
```diff
@@ -10,4 +10,5 @@ if grep -q docker /proc/1/cgroup; then
 fi
 
 # move kubeconfig to expected location
+mkdir -p ~/.kube
```
Comment (scothis) [inline clusters/kind/start.sh]: 👍

File: macros/invoke_core_deployer.sh [hunk 10]
```diff
@@ -4,6 +4,7 @@ name=$1
 curl_opts=$2
 expected_data=$3
 
+NAMESPACE=${NAMESPACE:-fats}
```
Comment (scothis) [inline macros/invoke_core_deployer.sh]: rm

NAMESPACE is part of the FATS contract. We should not define it here. The script should fail if the value is not set.

File: registries/docker-daemon/configure.sh [hunk 11]
```diff
@@ -3,15 +3,21 @@
 daemonConfig='/etc/docker/daemon.json'
 if test -f ${daemonConfig} && grep -q registry.kube-system.svc.cluster.local ${daemonConfig}; then
   echo "insecure registry previously configured"
-else
-  # Allow for insecure registries
-  sudo mkdir -p /etc/docker
-  echo '{ "insecure-registries": [ "registry.kube-system.svc.cluster.local:5000" ] }' | sudo tee ${daemonConfig} > /dev/null
-  sudo systemctl daemon-reload
-  sudo systemctl restart docker
+elif which systemctl > /dev/null; then
```
Comment (scothis) [inline registries/docker-daemon/configure.sh]: seems like a fragile test. What's you're goal here?

File: registries/docker-daemon/configure.sh [hunk 12]
```diff
@@ -3,15 +3,21 @@
...
+      echo '{}' | sudo tee ${daemonConfig} > /dev/null
+    fi
+    jq -s '.[0] * .[1]' <( cat ${daemonConfig}) <(echo '{ "insecure-registries": [ "registry.kube-system.svc.cluster.local:5000" ] }') | sudo tee ${daemonConfig} > /dev/null
+    sudo systemctl daemon-reload
+    sudo systemctl restart docker
+  fi
 fi
 
 IMAGE_REPOSITORY_PREFIX="registry.kube-system.svc.cluster.local:5000"
+CLUSTER_NAME=${CLUSTER_NAME-fats}
```
Comment (scothis) [inline registries/docker-daemon/configure.sh]: rm

CLUSTER_NAME is part of the FATS contract. We should not define it here. The script should fail if the value is not set.

File: registries/docker-daemon/start.sh [hunk 13]
```diff
@@ -5,7 +5,10 @@ echo "Installing a daemon registry"
 docker run -d -p 5000:5000 --name registry registry:2
 
 registry_ip=$(docker inspect --format "{{.NetworkSettings.IPAddress }}" registry)
-sudo su -c "echo \"${registry_ip} registry.kube-system.svc.cluster.local\" >> /etc/hosts"
+if ! grep registry /etc/hosts; then
```
Comment (scothis) [inline registries/docker-daemon/start.sh]: how about we remove any entry from hosts with `registry.kube-system.svc.cluster` and then add the new value?

File: .github/workflows/cleanup.sh (hunk 2 above)
Comment (dsyer) [reply .github/workflows/cleanup.sh]: Why not? It's useful for testing. Seems harmless.

File: .github/workflows/cleanup.sh (hunk 3 above)
Comment (dsyer) [reply .github/workflows/cleanup.sh]: Why not? It's useful for testing. Seems harmless.

File: .github/workflows/install.sh (hunk 4 above)
Comment (dsyer) [reply .github/workflows/install.sh]: Why not? It's useful for testing. Seems harmless.

File: .github/workflows/run.sh (hunk 5 above)
Comment (dsyer) [reply .github/workflows/run.sh]: Why not? It's useful for testing. Seems harmless.

File: registries/docker-daemon/configure.sh (hunk 11 above)
Comment (dsyer) [reply registries/docker-daemon/configure.sh]: To not fail if there is no systemctl (e.g. in a container).

File: registries/docker-daemon/start.sh (hunk 13 above)
Comment (dsyer) [reply registries/docker-daemon/start.sh]: You can't edit /etc/hosts in a container. This seemed like the least worst compromise. I guess we could attempt to edit it and fall back to this if it fails.

File: .github/workflows/cleanup.sh (hunk 2 above)
Comment (scothis) [reply .github/workflows/cleanup.sh]: this script is not part of FATS and may change at any time without notice. The harm is that the value is expected to be provided externally. By defaulting the value, we may not notice if the value isn't set, but should have been.

You can use these scripts as example for writing your own scripts that consume FATS, but these scripts should not be consumed directly. If you really want this behavior, copy the script and add it to your local copy.

File: macros/invoke_core_deployer.sh [hunk 14]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -o nounset
```
Comment (scothis) [inline macros/invoke_core_deployer.sh]: rm

macros are often sourced and should not modify the environment of the parent script

File: macros/invoke_knative_deployer.sh [hunk 15]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -o nounset
```
Comment (scothis) [inline macros/invoke_knative_deployer.sh]: rm

macros are often sourced and should not modify the environment of the parent script

File: registries/docker-daemon/configure.sh [hunk 16]
```diff
@@ -1,14 +1,21 @@
 #!/bin/bash
 
+set -o nounset
```
Comment (scothis) [inline registries/docker-daemon/configure.sh]: rm

this script is sourced and should not modify the environment of the parent script

File: registries/docker-daemon/configure.sh (hunk 11 above)
Comment (scothis) [reply registries/docker-daemon/configure.sh]: As proposed, the docker daemon won't know that the registry is insecure and any attempt to push or pull images from the registry will fail.

That logic is being implemented somewhere outside of this script. There should be some other signal that the registry is already configured. Whether systemctl is available is not an indication that the registry was configured.

File: registries/docker-daemon/start.sh (hunk 13 above)
Comment (scothis) [reply registries/docker-daemon/start.sh]: I'm open to finding a better way to define a one-off dns record. Until then, we can make the test more robust. The run will fail if the IP is not correct.

```suggestion
if ! grep registry.kube-system.svc.cluster.local /etc/hosts; then
```

File: registries/docker-daemon/configure.sh (hunk 11 above)
Comment (dsyer) [reply registries/docker-daemon/configure.sh]: It is, however, an indication that this script is going to fail. Everything works if the docker daemon is running on the host and you are in a container, but there is no need (or sense) then in trying to do anything with systemctl. We could use the /proc/1/cgroup test from the kind/start.sh script instead?

Comment (dsyer) [reply registries/docker-daemon/configure.sh]: See 6c45dea

File: registries/docker-daemon/configure.sh [hunk 17]
```diff
@@ -1,14 +1,19 @@
...
+    echo "insecure registry previously configured"
+  else
+    if ! test -f ${daemonConfig} || test -s ${daemonConfig}; then
+      sudo mkdir -p /etc/docker
+      echo '{}' | sudo tee ${daemonConfig} > /dev/null
+    fi
+    jq -s '.[0] * .[1]' <( cat ${daemonConfig}) <(echo '{ "insecure-registries": [ "registry.kube-system.svc.cluster.local:5000" ] }') | sudo tee ${daemonConfig} > /dev/null
+    sudo systemctl daemon-reload
+    sudo systemctl restart docker
+  fi
```
Comment (scothis) [inline registries/docker-daemon/configure.sh]:
```suggestion
  else
    # test that registry is configured correctly
    docker pull cloudfoundry/run:tiny \
      && docker tag cloudfoundry/run:tiny registry.kube-system.svc.cluster.local:5000/run:tiny \
      && docker push registry.kube-system.svc.cluster.local:5000/run:tiny
  fi
```

File: .github/workflows/run.sh [hunk 18]
```diff
@@ -10,6 +10,8 @@ source ${FATS_DIR}/.configure.sh
 ${FATS_DIR}/install.sh riff
 ${FATS_DIR}/install.sh kubectl
 
+FATS_REPO=${FATS_REPO-projectriff/fats}
+FATS_REFSPEC=${FATS_REFSPEC-master}
```
[the remark this answers is not in the document]
Comment (dsyer) [reply .github/workflows/run.sh]: I hate having env vars as part of the contract of a script. It doesn't matter to me that they would "normally" have. It increases the burden of inderstanding the code. However, if you insist (see c4285ec).

**Comment by dsyer** [comment]
> Please close this PR and open a new PR from inside the projectriff repo.

Why? Here you go anyway:

https://github.com/projectriff/fats/pull/240

**Comment by scothis** [comment]
>> Please close this PR and open a new PR from inside the projectriff repo.

> Why?

Same reason as before. CI isn't able to run on remote PRs since these tests fundamentally require credentials that are not available on remote PRs.

**Pull request closed without merging**